datafusion/spark/src/function/math/hex.rs

```diff
@@ -0,0 +1,404 @@
+        }
+        match &arg_types[0] {
+            DataType::Int64
+            | DataType::Utf8
+            | DataType::LargeUtf8
+            | DataType::Binary
+            | DataType::LargeBinary => Ok(vec![arg_types[0].clone()]),
+            DataType::Dictionary(key_type, value_type) => match value_type.as_ref() {
+                DataType::Int64
+                | DataType::Utf8
+                | DataType::LargeUtf8
+                | DataType::Binary
+                | DataType::LargeBinary => Ok(vec![arg_types[0].clone()]),
+                other => {
+                    if other.is_numeric() {
+                        Ok(vec![DataType::Dictionary(
+                            key_type.clone(),
+                            Box::new(DataType::Int64),
+                        )])
+                    } else {
+                        Err(unsupported_data_type_exec_err(
+                            "hex",
+                            "Numeric, String, or Binary",
+                            &arg_types[0],
+                        ))
+                    }
+                }
+            },
+            other => {
+                if other.is_numeric() {
+                    Ok(vec![DataType::Int64])
+                } else {
+                    Err(unsupported_data_type_exec_err(
+                        "hex",
+                        "Numeric, String, or Binary",
+                        &arg_types[0],
+                    ))
+                }
+            }
+        }
+    }
+}
+
+fn hex_int64(num: i64) -> String {
+    format!("{:X}", num)
+}
+
+#[inline(always)]
+fn hex_encode<T: AsRef<[u8]>>(data: T, lower_case: bool) -> String {
+    let mut s = String::with_capacity(data.as_ref().len() * 2);
+    if lower_case {
+        for b in data.as_ref() {
+            // Writing to a string never errors, so we can unwrap here.
+            write!(&mut s, "{b:02x}").unwrap();
+        }
+    } else {
+        for b in data.as_ref() {
+            // Writing to a string never errors, so we can unwrap here.
+            write!(&mut s, "{b:02X}").unwrap();
+        }
+    }
+    s
+}
+
+#[inline(always)]
+fn hex_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<String, std::fmt::Error> {
+    let hex_string = hex_encode(bytes, false);
+    Ok(hex_string)
+}
+
+/// Spark-compatible `hex` function
+pub fn spark_hex(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+    if args.len() != 1 {
+        return Err(DataFusionError::Internal(
+            "hex expects exactly one argument".to_string(),
+        ));
+    }
+
+    let input = match &args[0] {
+        ColumnarValue::Scalar(value) => ColumnarValue::Array(value.to_array()?),
+        ColumnarValue::Array(_) => args[0].clone(),
+    };
+
+    match &input {
+        ColumnarValue::Array(array) => match array.data_type() {
+            DataType::Int64 => {
+                let array = as_int64_array(array)?;
+
+                let hexed_array: StringArray =
```

Comment:
You can probably make this much faster by avoiding the allocation by using StringBuilder directly

If you are using StringArray something like this:
https://docs.rs/arrow/latest/arrow/array/type.GenericStringBuilder.html#example-incrementally-writing-strings-with-stdfmtwrite


If you are using StringView I think you'll have to provide a fast path for shorter strings and use the StringView builder for longer strings